modules/rtdModule/index.js

```diff
@@ -0,0 +1,208 @@
+  const dataWaitTimeout = setTimeout(() => {
+    processDone = true;
+    callback(dataReceived);
+  }, _moduleConfig.auctionDelay);
+
+  subModules.forEach(sm => {
+    sm.getData(adUnits, onDataReceived);
+  });
+
+  function onDataReceived(data) {
+    if (processDone) {
+      return
+    }
+    dataReceived.push(data);
+    if (dataReceived.length === callbackExpected) {
+      processDone = true;
+      clearTimeout(dataWaitTimeout);
+      callback(dataReceived);
+    }
+  }
+}
+
+/**
+ * run hook after bids request and before callback
+ * get data from provider and set key values to primary ad server
+ * @param {function} next - next hook function
+ * @param {AdUnit[]} adUnits received from auction
+ */
+export function setTargetsAfterRequestBids(next, adUnits) {
+  getProviderData(adUnits, (data) => {
+    if (data && Object.keys(data).length) {
+      const _mergedData = deepMerge(data);
+      if (Object.keys(_mergedData).length) {
+        setDataForPrimaryAdServer(_mergedData);
+      }
+    }
+    next(adUnits);
+  });
+}
+
+/**
+ * deep merge array of objects
+ * @param {array} arr - objects array
+ * @return {Object} merged object
+ */
+export function deepMerge(arr) {
+  if (!arr.length) {
+    return {};
+  }
+  return arr.reduce((merged, obj) => {
+    for (let key in obj) {
+      if (obj.hasOwnProperty(key)) {
+        if (!merged.hasOwnProperty(key)) merged[key] = obj[key];
+        else {
+          // duplicate key - merge values
+          const dp = obj[key];
+          for (let dk in dp) {
+            if (dp.hasOwnProperty(dk)) merged[key][dk] = dp[dk];
+          }
+        }
+      }
+    }
+    return merged;
+  }, {});
+}
+
+/**
+ * run hook before bids request
+ * get data from provider and set key values to primary ad server & bidders
+ * @param {function} fn - hook function
+ * @param {Object} reqBidsConfigObj - request bids object
+ */
+export function requestBidsHook(fn, reqBidsConfigObj) {
+  getProviderData(reqBidsConfigObj.adUnits || getGlobal().adUnits, (data) => {
+    if (data && Object.keys(data).length) {
+      const _mergedData = deepMerge(data);
+      if (Object.keys(_mergedData).length) {
+        setDataForPrimaryAdServer(_mergedData);
+        addIdDataToAdUnitBids(reqBidsConfigObj.adUnits || getGlobal().adUnits, _mergedData);
+      }
+    }
+    return fn.call(this, reqBidsConfigObj);
+  });
+}
+
+/**
+ * set data to primary ad server
+ * @param {Object} data - key values to set
+ */
+function setDataForPrimaryAdServer(data) {
+  if (!utils.isGptPubadsDefined()) {
+    utils.logError('window.googletag is not defined on the page');
+    return;
+  }
+  targeting.setTargetingForGPT(data, null);
+}
+
+/**
+ * @param {AdUnit[]} adUnits
+ *  @param {Object} data - key values to set
+ */
+function addIdDataToAdUnitBids(adUnits, data) {
+  adUnits.forEach(adUnit => {
+    adUnit.bids = adUnit.bids.map(bid => {
+      const rd = data[adUnit.code] || {};
+      return Object.assign(bid, rd);
```

Comment:
Should we be placing these inside some object on the bid rather than the top level of the bid?

Reply:
Yes, I'll place it inside "realTimeData" object.
Will be fixed on the next commit.